javalib/src/main/scala/java/io/BufferedInputStream.scala

```diff
@@ -1,129 +1,247 @@
 package java.io
 
-class BufferedInputStream(_in: InputStream, size: Int)
+/* A note on the implementation. IMO (coreyoconnor) neither the written or de-facto specification for
+ * BufferedInputStream are nicely designed. In particular the aspects:
+ *
+ * 1. protected fields
+ * 2. asynchronous close
+ *
+ * For async close: I cannot find an official spec on this, but evidence implies this is required:
+ * First are the existing implementations in Apache Harmony and OpenJDK. Second are the bug reports
+ * implying close must support async.
+ *
+ * - https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4225348
+ * - https://issues.apache.org/jira/browse/HARMONY-6014
+ *
+ * Full thread correctness seems near impossible, so the implementation only prioritizes async close.
+ *
+ * Mark support makes lenient use of the "might" and "may"s in the spec.
+ *
+ * One pattern is using `Option(in)` to capture a field to a local.
+ */
+class BufferedInputStream(_in: InputStream, initialSize: Int)
     extends FilterInputStream(_in)
     with Closeable
     with AutoCloseable {
 
-  if (size <= 0) throw new IllegalArgumentException("Buffer size <= 0")
+  if (initialSize <= 0) throw new IllegalArgumentException("Buffer size <= 0")
 
   def this(in: InputStream) = this(in, 8192)
 
+  // per spec close will release system resources. This implies buf should be set to null
+  // post close to ensure GC can release this resource
   /** The internal buffer array where the data is stored. */
-  protected[this] var buf = new Array[Byte](size)
+  protected[this] var buf = new Array[Byte](initialSize)
 
   /** The index one greater than the index of the last valid byte in the buffer. */
-  protected[this] var count = 0
+  protected[this] var count: Int = 0
 
-  /** The maximum read ahead allowed after a call to the mark method before subsequent calls to the reset method fail.. */
-  private[this] var markLimit = 0
+  /** The maximum read ahead allowed after a call to the mark method before subsequent calls to the
+    *  reset method fail..
+    */
+  protected[this] var marklimit: Int = 0
 
   /** The value of the pos field at the time the last mark method was called. */
-  private[this] var markpos = -1
+  protected[this] var markpos: Int = -1
 
   /** The current position in the buffer. */
-  protected[this] var pos = 0
-
-  private[this] var closed = false
-
-  /** The position of the last element in the buffer excluded */
-  private[this] var end = 0
-
-  override def available(): Int = {
-    if (closed) throw new IOException()
-    end - pos + in.available()
+  protected[this] var pos: Int = 0
+
+  override def available(): Int = Option(in) match {
```

Comment:
Well, in fact the above `ensureOpen()` is not good enough, because `read()` can reallocate a bigger `buf`, which means that it can concurrently set another non-null `buf` after the `buf = null`.

Therefore, I believe a dedicated
```scala
private[this] var closed: Boolean = false
```
and then do:
```scala
def ensureOpen(): (InputStream, Array[Byte]) = {
  /* First read `in` and `buf`, then `closed`. Since `closed` is the first thing
   * that is set to `true` in `close()`, we know that if `closed` is false, the
   * `in` and `buf` are non-null.
   */
  val in = this.in
  val buf = this.buf
  if (closed)
    throw new IOException("Operation on closed stream")
  (in, buf)
}
```
while `close()` starts by setting `closed` to `true`:
```scala
def close(): Unit = {
  if (!closed) {
    closed = true
    val in = this.in
    if (in != null)
      in.close()
    in = null
    buff = null
  }
}
```